test/integration/consul-container/test/ratelimit/ratelimit_test.go

```diff
@@ -0,0 +1,261 @@
+package ratelimit
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"github.com/testcontainers/testcontainers-go"
+
+	"github.com/hashicorp/consul/api"
+	libcluster "github.com/hashicorp/consul/test/integration/consul-container/libs/cluster"
+)
+
+const (
+	retryableErrorMsg    = "Unexpected response code: 429 (rate limit exceeded, try again later or against a different server)"
+	nonRetryableErrorMsg = "Unexpected response code: 503 (rate limit exceeded for operation that can only be performed by the leader, try again later)"
+)
+
+// TestRateLimit
+// This test validates
+// - enforcing mode
+//   - read_rate - returns 429 - was blocked and returns retryable error
+//   - write_rate - returns 503 - was blocked and is not retryable
+//   - on each
+//     - fires metrics forexceeding
+//     - logs for exceeding
+
+func TestRateLimit(t *testing.T) {
+	type action struct {
+		function           func(client *api.Client) error
+		rateLimitOperation string
+		rateLimitType      string // will become an array of strings
+	}
+	type operation struct {
+		action            action
+		expectedErrorMsg  string
+		expectExceededLog bool
+		expectBlockedLog  bool
+		expectMetric      bool
+	}
+	type testCase struct {
+		description string
+		cmd         string
+		operations  []operation
+	}
+
+	getKV := action{
+		function: func(client *api.Client) error {
+			_, _, err := client.KV().Get("foo", &api.QueryOptions{})
+			return err
+		},
+		rateLimitOperation: "KVS.Get",
+		rateLimitType:      "global/read",
+	}
+	putKV := action{
+		function: func(client *api.Client) error {
+			_, err := client.KV().Put(&api.KVPair{Key: "foo", Value: []byte("bar")}, &api.WriteOptions{})
+			return err
+		},
+		rateLimitOperation: "KVS.Apply",
+		rateLimitType:      "global/write",
+	}
+
+	testCases := []testCase{
+		// HTTP & net/RPC
+		{
+			description: "HTTP & net/RPC / Mode: disabled - errors: no / exceeded logs: no / blocked logs: no / metrics: no",
+			cmd:         "-hcl=limits { request_limits { mode = \"disabled\" read_rate = 0 write_rate = 0 }}",
```

Comment:
nit: we can avoid needing to escape those quotes by using backticks:

```suggestion
			cmd:         `-hcl=limits { request_limits { mode = "disabled" read_rate = 0 write_rate = 0 }}`,
```